MmSupervisorPkg/Core/MmSupervisorCore.c

```diff
@@ -169,126 +170,164 @@ PrepareCommonBuffers (
-      // But the memory itself is allocated under reserved..
-      mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].Type      = EfiRuntimeServicesData;
-      mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].Attribute = EFI_MEMORY_XP | EFI_MEMORY_SP;
-      if (CommRegionHob->MmCommonRegionType == MM_SUPERVISOR_BUFFER_T) {
-        Status = MmAllocateSupervisorPages (
-                   AllocateAnyPages,
-                   EfiRuntimeServicesData,
-                   CommRegionHob->MmCommonRegionPages,
-                   (EFI_PHYSICAL_ADDRESS *)&mInternalCommBufferCopy[CommRegionHob->MmCommonRegionType]
-                   );
-      } else {
-        Status = MmAllocatePages (
-                   AllocateAnyPages,
-                   EfiRuntimeServicesData,
-                   CommRegionHob->MmCommonRegionPages,
-                   (EFI_PHYSICAL_ADDRESS *)&mInternalCommBufferCopy[CommRegionHob->MmCommonRegionType]
-                   );
-      }
-
-      ASSERT_EFI_ERROR (Status);
-      if (EFI_ERROR (Status)) {
-        DEBUG ((DEBUG_ERROR, "%a - Failed to allocate internal buffer copy, please consider adjust TSEG size... - %r\n", __FUNCTION__, Status));
-        goto Exit;
-      }
+  CommRegionHob = GET_GUID_HOB_DATA (GuidHob.Guid);
+  if (CommRegionHob->MmCommonRegionType == MM_SUPERVISOR_BUFFER_T) {
+    if (mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].PhysicalStart != 0) {
+      DEBUG ((DEBUG_ERROR, "%a - Duplicated hobs for type %x!!\n", __FUNCTION__, CommRegionHob->MmCommonRegionType));
+      Status = EFI_ALREADY_STARTED;
+      goto Exit;
+    }
 
-      mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].VirtualStart = 0;
+    if (!MmIsBufferOutsideMmValid (
+            mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].PhysicalStart,
+            EFI_PAGES_TO_SIZE (mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].NumberOfPages)
+            ))
+    {
       DEBUG ((
-        DEBUG_INFO,
-        "%a - Populating MM Access Buffer Type %d to 0x%p with 0x%x pages\n",
+        DEBUG_ERROR,
+        "%a - Buffer (%p) invalid for type %x!!\n",
         __FUNCTION__,
-        CommRegionHob->MmCommonRegionType,
         mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].PhysicalStart,
-        mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].NumberOfPages
+        CommRegionHob->MmCommonRegionType
         ));
+      Status = EFI_BAD_BUFFER_SIZE;
+      ASSERT (FALSE);
+      goto Exit;
     }
 
-    GuidHob.Guid = GET_NEXT_HOB (GuidHob);
-    GuidHob.Guid = GetNextGuidHob (&gMmCommonRegionHobGuid, GuidHob.Guid);
+    mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].PhysicalStart = CommRegionHob->MmCommonRegionAddr;
+    mMmSupervisorAccessBuffer[CommRegionHob->MmCommonRegionType].NumberOfPages = CommRegionHob->MmCommonRegionPages;
+    // But the memory itself is allocated under reserved..

```

Comment:
I'm confused by this comment. Is this trying to say on the PEI/DXE side this is allocated as reserved but here we are doing runtime services data?

Reply:
The allocations are coming from the PEI phase, and then then internal buffer is allocated in MM, removed the comment.